aidial_adapter_vertexai/dial_api/embedding_inputs.py

```diff
@@ -0,0 +1,116 @@
+from typing import (
+    Any,
+    AsyncIterator,
+    Callable,
+    Coroutine,
+    List,
+    TypeVar,
+    assert_never,
+    cast,
+)
+
+from aidial_sdk.chat_completion.request import Attachment
+from aidial_sdk.embeddings.request import EmbeddingsRequest
+
+from aidial_adapter_vertexai.chat.errors import ValidationError
+
+T = TypeVar("T")
+
+Coro = Coroutine[T, Any, Any]
+Tokens = List[int]
+
+
+async def reject_tokens(tokens: Tokens):
+    raise ValidationError(
+        "Tokens in the input are not supported, provide text instead. "
+        "When Langchain AzureOpenAIEmbeddings class is used, set 'check_embedding_ctx_length=False' to disable tokenization."
+    )
+
+
+EMPTY_INPUT_LIST_ERROR = ValidationError(
+    "Empty list in an element of custom_input list"
+)
+
+ATTACHMENT_ERROR = ValidationError("Attachments are not supported")
+
+
+async def reject_attachment(attachment: Attachment):
+    raise ATTACHMENT_ERROR
+
+
+async def collect_embedding_inputs(
+    request: EmbeddingsRequest,
+    *,
+    on_text: Callable[[str], Coro[T]],
+    on_tokens: Callable[[Tokens], Coro[T]] = reject_tokens,
+    on_attachment: Callable[[Attachment], Coro[T]] = reject_attachment,
+    on_mixed: Callable[[List[str | Attachment]], Coro[T]],
+) -> AsyncIterator[T]:
+
+    async def on_text_or_attachment(input: str | Attachment) -> T:
+        if isinstance(input, str):
+            return await on_text(input)
+        else:
+            return await on_attachment(input)
+
+    if isinstance(request.input, str):
+        yield await on_text(request.input)
+    elif isinstance(request.input, list):
+
+        is_list_of_tokens = False
+        for input in request.input:
+            if isinstance(input, str):
+                yield await on_text(input)
+            elif isinstance(input, list):
+                yield await on_tokens(input)
+            else:
+                is_list_of_tokens = True
+
+        if is_list_of_tokens:
+            yield await on_tokens(cast(Tokens, request.input))
```

Comment:
If I understand correctly, we're currently not supporting Tokens at all

But if we do, I think this part will just cause some bugs

Imagine input like that 
```python
bad_input = [[1,2], [2, 'test'], 'test_string']
```

This request will be processed as valid, and `on_tokens` would receive `[2, 'test']` first, and then this whole list


---
I propose to make something like this

```python
Tokens = List[int]
def _is_valid_tokens(input: Any) -> bool:
    return isinstance(input, list) and all([isinstance(value, int) for value in input])

...

if isinstance(request.input, str):
    ...
elif _is_valid_tokens(request.input):
    yield await on_tokens(cast(Tokens, request.input))
elif isistance(request.input, list):
        for input in request.input:
            if isinstance(input, str):
                yield await on_text(input)
            elif _is_valid_tokens(input, list):
                yield await on_tokens(input)
            else:
                assert_never(input)
```

What do you think?

Reply:
That's a valid point, however we will redo the same work that Pydantic already doing for us, since it validates objects when they are constructed. 

The following code fails with a runtime error:

```python
from aidial_sdk.embeddings.request import EmbeddingsRequest

EmbeddingsRequest(input=[[1,2], [2, 'test'], 'test_string'])
```